File: src/main/java/com/wesker926/leetcode/algorithms/p1610/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/16
// @description: 坐标转换 + 滑动窗口
package s1

import (
	"math"
	"sort"
)

func visiblePoints(points [][]int, angle int, location []int) int {
	same, polars := 0, make([]float64, 0, len(points)*2)
	for _, point := range points {
		if point[0] == location[0] && point[1] == location[1] {
			same++
		} else {
			polars = append(polars, math.Atan2(float64(point[1]-location[1]), float64(point[0]-location[0])))
		}
	}

	n := len(polars)
	sort.Slice(polars, func(i, j int) bool {
		return polars[i] < polars[j]
	})
	for i := 0; i < n; i++ {
		polars = append(polars, polars[i]+2*math.Pi)
	}

	scope, max := float64(angle)/180*math.Pi, 0
	for i, j := 0, 0; i < n; i++ {
		for ; j < len(polars) && polars[j] <= polars[i]+scope; j++ {
		}
		if j-i > max {
			max = j - i
		}
	}
	return max + same
}

File: src/main/java/com/wesker926/leetcode/algorithms/p1610/s1/Solution.java
package com.wesker926.leetcode.algorithms.p1610.s1;

import java.util.Arrays;
import java.util.List;

/**
 * @author wesker.gh
 * @date 2021/12/16
 * @description 坐标转换 + 滑动窗口
 */
public class Solution {
    public int visiblePoints(List<List<Integer>> points, int angle, List<Integer> location) {
        int same = 0, x1 = location.get(0), x2, y1 = location.get(1), y2, idx = 0, n = points.size();
        double[] polars = new double[n * 2];
        for (List<Integer> point : points) {
            x2 = point.get(0);
            y2 = point.get(1);
            if (x1 == x2 && y1 == y2) {
                same++;
            } else {
                polars[idx++] = Math.atan2(y2 - y1, x2 - x1);
            }
        }

        n = idx;
        Arrays.sort(polars, 0, n);
        for (idx = 0; idx < n; idx++) {
            polars[idx + n] = polars[idx] + Math.PI * 2;
        }

        int max = 0;
        double scope = angle * Math.PI / 180;
        for (int i = 0, j = 0; i < n; i++) {
            for (; j < polars.length && polars[j] <= scope + polars[i]; j++) {
            }
            max = Math.max(max, j - i);
        }
        return max + same;
    }
}
